source/manual/conventions-in-rails-applications.html.md

```diff
@@ -0,0 +1,301 @@
+[slimmer]: https://github.com/alphagov/slimmer
+[issue-example]: https://github.com/alphagov/govuk_app_config/issues/121
+
+## Data storage
+
+For non-specialist database needs you should use
+[PostgreSQL](https://www.postgresql.org/) with [ActiveRecord][]. The
+[db/seeds.rb][] can be used to populate the database for development and test
+environments.
+
+For [key-value datastore][] access (such as queues or distributed caches)
+[Redis][] should be used.
+
+For file storage local to an application [Amazon S3][] is the preferred choice,
+where this needs to be associated with a database [ActiveStorage][] should be
+used.
+
+[ActiveRecord]: https://guides.rubyonrails.org/active_record_basics.html
+[db/seeds.rb]: https://github.com/alphagov/content-publisher/blob/master/db/seeds.rb
+[key-value datastore]: https://en.wikipedia.org/wiki/Key-value_database
+[Redis]: https://redis.io/
+[Amazon S3]: https://aws.amazon.com/s3/
+[ActiveStorage]: https://guides.rubyonrails.org/active_storage_overview.html
+
+## Background job processing
+
+The preferred approach for background job processing is to use [ActiveJob][]
+with [Sidekiq][] with the [govuk_sidekiq][] gem providing configuration.
+ActiveJob is preferred due to it's closer integration of Rails components
+(such as [ActionMailer][]).
+
+Scheduled background jobs for applications should also make use of Sidekiq,
+where [sidekiq-scheduler][] is the conventional choice to achieve this.
+
+[Sidekiq]: https://sidekiq.org
+[ActiveJob]: https://guides.rubyonrails.org/active_job_basics.html
+[ActionMailer]: https://guides.rubyonrails.org/action_mailer_basics.html
+[sidekiq-scheduler]: https://github.com/moove-it/sidekiq-scheduler
+
+## JavaScript package management
+
+As per Rails conventions GOV.UK applications should use
+[Yarn](https://yarnpkg.com) for JavaScript packages. This is because it
+integrates better with Rails than [NPM](https://www.npmjs.com/).
+
+## Frontend assets
+
+GOV.UK applications use the Rails [asset pipeline][] for building assets. We
+aspire to migrate to [webpacker][] as the conventional approach when we have
+resolved [blocking issues][].
+
+[asset pipeline]: https://guides.rubyonrails.org/asset_pipeline.html
+[webpacker]: https://github.com/rails/webpacker
+[blocking issues]: https://github.com/alphagov/govuk_publishing_components/issues/505
+
+## Sending emails
+
+[GOV.UK Notify][] is the preferred approach for sending email. The
+[mail-notify][] gem is conventionally used to integrate Notify with
+[ActionMailer][]. There is a shared GOV.UK Publishing Notify account that can
+be used across GOV.UK publishing/admin applications.
+
+[GOV.UK Notify]: https://www.notifications.service.gov.uk/
+[mail-notify]: https://github.com/dxw/mail-notify
+
+## Configure Rails as `api_only` for API projects
+
+For applications that do not serve requests to a web browser you should
+configure the Rails application to be [API only][].
+
+[API only]: https://guides.rubyonrails.org/api_app.html
```

Comment:
Maybe not for this PR - but would be useful to have guidance around serving both an API and HTML from the same application - (is that even a thing we should be doing?)

Reply:
what is our preferred why to serialise JSON - jbuilder, AMS etc...

Reply:
> Maybe not for this PR - but would be useful to have guidance around serving both an API and HTML from the same application - (is that even a thing we should be doing?)

I guess if you've got a HTML rendering Rails app it's ok to have a few API endpoints, nearly all apps implement this through /healthcheck endpoints. Though too many might be a sign you've got a monolith growing (of course, Whitehall has quite a rich API)

Where I think this smells though is when an application that is essentially API only has HTML endpoints. We have this problem with Publishing API, where we have one, seldomly used, [debug endpoint](https://github.com/alphagov/publishing-api/blob/master/app/controllers/debug_controller.rb) that is the lone requirement for all the extra Rails baggage needed for view rendering etc.





> and what is our preferred way to serialise JSON - jbuilder, AMS etc...

I don't think we've managed to standardise on that, Whitehall uses jbuilder but I don't think anything else does. Instead it seems to get wrapped up in the ambiguous concept of Presenters where in some apps they're used for serializing data and others as a presentation model. It'd be great if our next API driven app could establish precedents for serialisation to move us away from mixing them in with the presenter concept.

Reply:
Yup agree - I think overall we haven't got any formalisation on guidance around building APIs in Rails. Sorry for opening that can of worms - probably a conversation to have in another place, as it doesn't seem helpful for this PR. I can't think of anything else to add to this section for now - so happy to resolve this convo!

Reply:
Gotcha, yes I don't think we've got any clear conventions in the direction of adding an API to an app. We have conventions the other way: for example, Email Alert API is a machine only API and we have a different app, Email Alert Frontend, to provide the users interface. 

I guess though when you hit the situation like you two did with signon, where the system is architectured as a public facing app, then you're pretty snookered to add an API on later without doing it directly to the app. So I wonder if this is more broader architectural convention of creating systems API first. Definitely think this ventures outside of this scope, but good discussion none the less 👍